File: backend/src/Blackbaud.Hackathon.Platform.Service/BusinessLogic/NotificationService.cs
using Microsoft.AspNetCore.SignalR;
using Blackbaud.Hackathon.Platform.Service.Hubs;

namespace Blackbaud.Hackathon.Platform.Service.BusinessLogic;

/// <summary>
/// Implementation of notification service using SignalR
/// Sends real-time notifications to connected clients
/// </summary>
public class NotificationService : INotificationService
{
    private readonly IHubContext<NotificationHub> _hubContext;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(IHubContext<NotificationHub> hubContext, ILogger<NotificationService> logger)
    {
        _hubContext = hubContext;
        _logger = logger;
    }

    /// <summary>
    /// Notify all users in a hackathon that idea was submitted
    /// </summary>
    public async Task NotifyIdeaSubmitted(int hackathonId, object ideaData)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("IdeaSubmitted", new
                {
                    idea = ideaData,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified hackathon {hackathonId} of idea submission");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying idea submission for hackathon {hackathonId}");
        }
    }

    /// <summary>
    /// Notify that comment was added to idea
    /// </summary>
    public async Task NotifyCommentAdded(int hackathonId, int ideaId, object commentData)
    {
        try
        {
            var notification = new
            {
                ideaId = ideaId,
                comment = commentData,
                timestamp = DateTime.UtcNow
            };

            // Notify hackathon group
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("CommentAdded", notification);

            // Notify idea watchers
            await _hubContext.Clients
                .Group($"idea_{ideaId}")
                .SendAsync("CommentAdded", notification);

            _logger.LogInformation($"Notified of comment on idea {ideaId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying comment for idea {ideaId}");
        }
    }

    /// <summary>
    /// Notify that rating was submitted for idea
    /// </summary>
    public async Task NotifyRatingSubmitted(int hackathonId, int ideaId, object ratingData)
    {
        try
        {
            var notification = new
            {
                ideaId = ideaId,
                rating = ratingData,
                timestamp = DateTime.UtcNow
            };

            // Notify judges
            await _hubContext.Clients
                .Group($"judging_{hackathonId}")
                .SendAsync("RatingSubmitted", notification);

            // Notify idea watchers
            await _hubContext.Clients
                .Group($"idea_{ideaId}")
                .SendAsync("RatingUpdated", notification);

            _logger.LogInformation($"Notified of rating submission for idea {ideaId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying rating for idea {ideaId}");
        }
    }

    /// <summary>
    /// Notify that idea was deleted
    /// </summary>
    public async Task NotifyIdeaDeleted(int hackathonId, int ideaId)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("IdeaDeleted", new
                {
                    ideaId = ideaId,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified of idea {ideaId} deletion");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying idea deletion {ideaId}");
        }
    }

    /// <summary>
    /// Notify that idea status changed
    /// </summary>
    public async Task NotifyIdeaStatusChanged(int hackathonId, int ideaId, string newStatus)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("IdeaStatusChanged", new
                {
                    ideaId = ideaId,
                    status = newStatus,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified of status change for idea {ideaId}: {newStatus}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying status change for idea {ideaId}");
        }
    }

    /// <summary>
    /// Notify that judging deadline is approaching
    /// </summary>
    public async Task NotifyJudgingDeadline(int hackathonId, DateTime deadline)
    {
        try
        {
            await _hubContext.Clients
                .Group($"judging_{hackathonId}")
                .SendAsync("JudgingDeadlineApproaching", new
                {
                    deadline = deadline,
                    hoursRemaining = (deadline - DateTime.UtcNow).TotalHours,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified judges of deadline for hackathon {hackathonId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying judging deadline for hackathon {hackathonId}");
        }
    }

    /// <summary>
    /// Notify that winner was announced
    /// </summary>
    public async Task NotifyWinnerAnnounced(int hackathonId, object winnerData)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("WinnerAnnounced", new
                {
                    winner = winnerData,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified of winner announcement for hackathon {hackathonId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying winner announcement for hackathon {hackathonId}");
        }
    }

    /// <summary>
    /// Notify team member joined
    /// </summary>
    public async Task NotifyTeamMemberJoined(int hackathonId, int teamId, object memberData)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("TeamMemberJoined", new
                {
                    teamId = teamId,
                    member = memberData,
                    timestamp = DateTime.UtcNow
                });

            await _hubContext.Clients
                .Group($"team_{teamId}")
                .SendAsync("TeamMemberJoined", new
                {
                    teamId = teamId,
                    member = memberData,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified of member joined team {teamId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying team member join for team {teamId}");
        }
    }

    /// <summary>
    /// Notify team member left
    /// </summary>
    public async Task NotifyTeamMemberLeft(int hackathonId, int teamId, string userId)
    {
        try
        {
            await _hubContext.Clients
                .Group($"hackathon_{hackathonId}")
                .SendAsync("TeamMemberLeft", new
                {
                    teamId = teamId,
                    userId = userId,
                    timestamp = DateTime.UtcNow
                });

            await _hubContext.Clients
                .Group($"team_{teamId}")
                .SendAsync("TeamMemberLeft", new
                {
                    teamId = teamId,
                    userId = userId,
                    timestamp = DateTime.UtcNow
                });

            _logger.LogInformation($"Notified of member left team {teamId}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, $"Error notifying team member leave for team {teamId}");
        }
    }
}

File: frontend/src/app/services/notification.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { HubConnection, HubConnectionBuilder, HubConnectionState } from '@microsoft/signalr';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { environment } from '../../environments/environment';

export interface Notification {
  id?: string;
  type: 'comment' | 'rating' | 'idea' | 'team' | 'judge' | 'winner' | 'deadline' | 'status';
  title: string;
  message: string;
  data?: any;
  timestamp: Date;
  read?: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService implements OnDestroy {
  private hubConnection: HubConnection;
  private destroy$ = new Subject<void>();

  // Notification streams
  notifications$ = new BehaviorSubject<Notification[]>([]);
  commentAdded$ = new Subject<any>();
  ratingSubmitted$ = new Subject<any>();
  ideaSubmitted$ = new Subject<any>();
  ideaDeleted$ = new Subject<any>();
  ideaStatusChanged$ = new Subject<any>();
  judgeOnline$ = new Subject<any>();
  judgeOffline$ = new Subject<any>();
  teamMemberJoined$ = new Subject<any>();
  teamMemberLeft$ = new Subject<any>();
  winnerAnnounced$ = new Subject<any>();
  judgingDeadline$ = new Subject<any>();

  // Connection state
  isConnected$ = new BehaviorSubject<boolean>(false);
  unreadCount$ = new BehaviorSubject<number>(0);
  onlineJudges$ = new BehaviorSubject<any[]>([]);

  constructor() {
    const hubUrl = `${environment.apiUrl.replace('api/', '')}hubs/notifications`;
    
    this.hubConnection = new HubConnectionBuilder()
      .withUrl(hubUrl, {
        accessTokenFactory: () => this.getToken(),
        withCredentials: true
      })
      .withAutomaticReconnect([0, 2000, 5000, 10000, 30000])
      .withServerTimeout(1000 * 60 * 10) // 10 minutes timeout
      .build();

    this.setupHubEventListeners();
  }

  /**
   * Start the SignalR connection
   */
  start(): Promise<void> {
    if (this.hubConnection.state === HubConnectionState.Connected) {
      return Promise.resolve();
    }

    return this.hubConnection.start()
      .then(() => {
        console.log('SignalR connected');
        this.isConnected$.next(true);
      })
      .catch((err: unknown) => {
        console.error('SignalR connection error:', err);
        this.isConnected$.next(false);
        return Promise.reject(err);
      });
  }

  /**
   * Stop the SignalR connection
   */
  stop(): Promise<void> {
    return this.hubConnection.stop()
      .then(() => {
        this.isConnected$.next(false);
      })
      .catch((err: unknown) => {
        console.error('SignalR stop error:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Join a hackathon group for notifications
   */
  joinHackathon(hackathonId: number): Promise<void> {
    return this.hubConnection.invoke('JoinHackathon', hackathonId)
      .catch((err: unknown) => {
        console.error('Failed to join hackathon group:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Leave a hackathon group
   */
  leaveHackathon(hackathonId: number): Promise<void> {
    return this.hubConnection.invoke('LeaveHackathon', hackathonId)
      .catch((err: unknown) => {
        console.error('Failed to leave hackathon group:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Join judging session
   */
  joinJudging(hackathonId: number): Promise<void> {
    return this.hubConnection.invoke('JoinJudging', hackathonId)
      .catch((err: unknown) => {
        console.error('Failed to join judging:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Leave judging session
   */
  leaveJudging(hackathonId: number): Promise<void> {
    return this.hubConnection.invoke('LeaveJudging', hackathonId)
      .catch((err: unknown) => {
        console.error('Failed to leave judging:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Watch specific idea for updates
   */
  watchIdea(ideaId: number): Promise<void> {
    return this.hubConnection.invoke('WatchIdea', ideaId)
      .catch((err: unknown) => {
        console.error('Failed to watch idea:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Stop watching specific idea
   */
  unwatchIdea(ideaId: number): Promise<void> {
    return this.hubConnection.invoke('UnwatchIdea', ideaId)
      .catch((err: unknown) => {
        console.error('Failed to unwatch idea:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Get list of online judges
   */
  getOnlineJudges(hackathonId: number): Promise<any> {
    return this.hubConnection.invoke('GetOnlineJudges', hackathonId)
      .catch((err: unknown) => {
        console.error('Failed to get online judges:', err);
        return Promise.reject(err);
      });
  }

  /**
   * Add notification to list
   */
  private addNotification(notification: Notification): void {
    const newNotification: Notification = {
      ...notification,
      id: `${notification.type}-${Date.now()}`,
      read: false
    };

    const current = this.notifications$.value;
    this.notifications$.next([newNotification, ...current]);
    this.unreadCount$.next(current.length + 1);

    // Auto-remove notification after 10 seconds
    setTimeout(() => {
      const updated = this.notifications$.value.filter(n => n.id !== newNotification.id);
      this.notifications$.next(updated);
    }, 10000);
  }

  /**
   * Mark notification as read
   */
  markAsRead(notificationId: string): void {
    const updated = this.notifications$.value.map(n => 
      n.id === notificationId ? { ...n, read: true } : n
    );
    this.notifications$.next(updated);
    const unread = updated.filter(n => !n.read).length;
    this.unreadCount$.next(unread);
  }

  /**
   * Clear all notifications
   */
  clearAll(): void {
    this.notifications$.next([]);
    this.unreadCount$.next(0);
  }

  /**
   * Setup all hub event listeners
   */
  private setupHubEventListeners(): void {
    // Comment added
    this.hubConnection.on('CommentAdded', (data: any) => {
      this.addNotification({
        type: 'comment',
        title: 'New Comment',
        message: `${data.comment.authorName} commented on an idea`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.commentAdded$.next(data);
    });

    // Rating submitted
    this.hubConnection.on('RatingSubmitted', (data: any) => {
      this.addNotification({
        type: 'rating',
        title: 'New Rating',
        message: `${data.rating.judgeEmail} rated an idea`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.ratingSubmitted$.next(data);
    });

    // Rating updated
    this.hubConnection.on('RatingUpdated', (data: any) => {
      this.ratingSubmitted$.next(data);
    });

    // Idea submitted
    this.hubConnection.on('IdeaSubmitted', (data: any) => {
      this.addNotification({
        type: 'idea',
        title: 'New Idea',
        message: `${data.idea.authorName} submitted: "${data.idea.title}"`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.ideaSubmitted$.next(data);
    });

    // Idea deleted
    this.hubConnection.on('IdeaDeleted', (data: any) => {
      this.addNotification({
        type: 'idea',
        title: 'Idea Deleted',
        message: `An idea has been deleted`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.ideaDeleted$.next(data);
    });

    // Idea status changed
    this.hubConnection.on('IdeaStatusChanged', (data: any) => {
      this.addNotification({
        type: 'status',
        title: 'Status Changed',
        message: `An idea status changed to ${data.status}`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.ideaStatusChanged$.next(data);
    });

    // Judge came online
    this.hubConnection.on('JudgeOnline', (data: any) => {
      this.judgeOnline$.next(data);
    });

    // Judge went offline
    this.hubConnection.on('JudgeOffline', (data: any) => {
      this.judgeOffline$.next(data);
    });

    // Judge status changed
    this.hubConnection.on('JudgeStatusChanged', (data: any) => {
      if (data.isOnline) {
        this.judgeOnline$.next(data);
      } else {
        this.judgeOffline$.next(data);
      }
    });

    // Team member joined
    this.hubConnection.on('TeamMemberJoined', (data: any) => {
      this.addNotification({
        type: 'team',
        title: 'Team Member Joined',
        message: `${data.member.email} joined a team`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.teamMemberJoined$.next(data);
    });

    // Team member left
    this.hubConnection.on('TeamMemberLeft', (data: any) => {
      this.addNotification({
        type: 'team',
        title: 'Team Member Left',
        message: `A team member left`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.teamMemberLeft$.next(data);
    });

    // Winner announced
    this.hubConnection.on('WinnerAnnounced', (data: any) => {
      this.addNotification({
        type: 'winner',
        title: 'Winner Announced!',
        message: `Congratulations to "${data.winner.title}"!`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.winnerAnnounced$.next(data);
    });

    // Judging deadline approaching
    this.hubConnection.on('JudgingDeadlineApproaching', (data: any) => {
      const hours = Math.round(data.hoursRemaining);
      this.addNotification({
        type: 'deadline',
        title: 'Deadline Approaching',
        message: `Judging deadline in ${hours} hours!`,
        data,
        timestamp: new Date(data.timestamp)
      });
      this.judgingDeadline$.next(data);
    });

    // Online judges list
    this.hubConnection.on('OnlineJudgesList', (data: any) => {
      this.onlineJudges$.next(data.judges || []);
    });

    // Connection events
    this.hubConnection.onreconnecting(() => {
      console.log('SignalR reconnecting...');
      this.isConnected$.next(false);
    });

    this.hubConnection.onreconnected(() => {
      console.log('SignalR reconnected');
      this.isConnected$.next(true);
    });

    this.hubConnection.onclose(() => {
      console.log('SignalR disconnected');
      this.isConnected$.next(false);
    });
  }

  /**
   * Get JWT token from localStorage
   */
  private getToken(): string {
    return localStorage.getItem('access_token') || '';
  }

  /**
   * Cleanup on service destroy
   */
  ngOnDestroy(): void {
    this.destroy$.next();
    this.destroy$.complete();
    this.stop().catch(() => {});
  }
}
